Source/Tools/LinuxEmulation/LinuxSyscalls/Syscalls.cpp

```diff
@@ -393,21 +393,43 @@ struct StackFramePlusRet {
   uint64_t Pad;
 };
 
+#ifdef _M_ARM_64
+#define ExitWithoutStack(Result) \
+  __asm volatile( \
+    "mov x8, %[SyscallNum];" \
+    "mov x0, %[Result];" \
+    "svc #0;" \
+    :: [SyscallNum] "i" (SYSCALL_DEF(exit)) \
+    , [Result] "r" (Result) \
+    : "memory", "x0", "x8");
+#else
+#define ExitWithoutStack(Result) \
+  __asm volatile( \
+    "mov %[Result], %%rdi;" \
+    "syscall;" \
+    :: "a" (SYSCALL_DEF(exit)) \
+    , [Result] "r" (Result) \
+    : "memory", "rdi");
+#endif
 [[noreturn]]
 static void Clone3HandlerRet() {
   StackFrameData *Data = (StackFrameData*)alloca(0);
```

Comment:
Since these functions have become increasingly complex, do we actually need Clone2HandlerRet and Clone3HandlerRet to be separate functions? If we change the `arg` parameter to be e.g. `std::optional<void*>`, we can just use `StackFrameData* Data = (StackFrameData*)(arg ? arg : alloca(0)) and keep the rest of the functions shared.

The ExitWithoutStack thing doesn't need to be a separate macro/function either then, since it's just used once.

Reply:
Implementations merged.